target/linux/ramips/dts/mt7628an_netgear_r6020.dts

```diff
@@ -0,0 +1,175 @@
+// SPDX-License-Identifier: GPL-2.0-or-later OR MIT
+/dts-v1/;
+
+#include "mt7628an.dtsi"
+
+#include <dt-bindings/gpio/gpio.h>
+#include <dt-bindings/input/input.h>
+
+/ {
+	compatible = "netgear,r6020", "mediatek,mt7628an-soc";
+	model = "Netgear R6020";
+
+	aliases {
+		led-boot = &led_power;
+		led-failsafe = &led_power;
+		led-running = &led_power;
+		led-upgrade = &led_power;
+		label-mac-device = &ethernet;
+	};
+
+	keys {
+		compatible = "gpio-keys";
+
+		reset {
+			label = "reset";
+			gpios = <&gpio 38 GPIO_ACTIVE_LOW>;
+			linux,code = <KEY_RESTART>;
+		};
+	};
+
+	leds {
+		compatible = "gpio-leds";
+
+		lan {
+			label = "r6020:green:lan";
+			gpios = <&gpio 44 GPIO_ACTIVE_LOW>;
+		};
+
+		led_power: power {
+			label = "r6020:green:power";
+			gpios = <&gpio 43 GPIO_ACTIVE_LOW>;
+		};
+
+		wlan2g_green {
+			label = "r6020:green:wlan2g";
+			gpios = <&gpio 42 GPIO_ACTIVE_LOW>;
+			linux,default-trigger = "phy0tpt";
+		};
+
+		wlan2g_orange {
+			label = "r6020:orange:wlan2g";
+			gpios = <&gpio 41 GPIO_ACTIVE_LOW>;
+		};
+
+		wan_green {
+			label = "r6020:green:wan";
+			gpios = <&gpio 40 GPIO_ACTIVE_LOW>;
+		};
+
+		wan_orange {
+			label = "r6020:orange:wan";
+			gpios = <&gpio 39 GPIO_ACTIVE_LOW>;
+		};
+	};
+};
+
+&state_default {
+	gpio {
+		groups = "p0led_an", "p1led_an", "p2led_an",
+			       "p3led_an", "p4led_an", "wdt",
+			       "wled_an", "uart1";
+		function = "gpio";
+	};
+};
+
+&spi0 {
+	status = "okay";
+
+	flash@0 {
+		compatible = "jedec,spi-nor";
+		reg = <0>;
+		spi-max-frequency = <86000000>;
+		m25p,fast-read;
+		
+		partitions {
+			compatible = "fixed-partitions";
+			#address-cells = <1>;
+			#size-cells = <1>;
+
+			partition@0 {
+				label = "u-boot";
+				reg = <0x0 0x40000>;
+				read-only;
+			};
+
+			factory: partition@40000 {
+				label = "factory";
+				reg = <0x40000 0x20000>;
+				read-only;
+			};
+
+			nvram: partition@60000 {
+				label = "nvram";
+				reg = <0x60000 0x30000>;
+				read-only;
+			};
+
+			partition@90000 {
+				compatible = "denx,uimage";
+				label = "firmware";
+				reg = <0x90000 0x6f0000>;
+			};
+
+			partition@780000 {
+				label = "ML";
+				reg = <0x780000 0x20000>;
+				read-only;
+			};
+
+			partition@7a0000 {
+				label = "ML1";
+				reg = <0x7a0000 0x20000>;
+				read-only;
+			};
+
+			partition@7c0000 {
+				label = "ML2";
+				reg = <0x7c0000 0x20000>;
+				read-only;
+			};
+		
+			partition@7e0000 {
+				label = "POT";
+				reg = <0x7e0000 0x10000>;
+				read-only;
+			};
+
+			partition@7f0000 {
+				label = "reserved";
+				reg = <0x7f0000 0x10000>;
+				read-only;
+			};
+		};
+	};
+};
+
+&wmac {
+	status = "okay";
+	mtd-mac-address = <&nvram 0x100b0>;
+	mediatek,mtd-eeprom = <&factory 0x20000>;
+};
+
+&ethernet {
+	mtd-mac-address = <&factory 0x8004>;
```

Comment:
Nothing in 0x28, 0x2e, 0xe000, 0xe006 or somewhere else in factory partition?

Reply:
It's possible it was corrupted when I got to it but this is the address where I found the MAC that's on the sticker.

Reply:
Yes, because that's the location of the 5g MAC address in calibration data. So, now the questions remain 1. whether this MAC address is supposed to be used for LAN as well and 2. whether the appropriate address is also found in one of the locations named above that are designated for the ethernet address on mt76xx platforms.

Reply:
The factory section was definitely corrupted.  The address used by the R6080, R6120, etc is the correct location of the LAN MAC.

Reply:
You won't convince me by just stating that you are right. You will convince me by reporting the content of 0x4, 0x8004, 0x28, 0x2e, 0xe000, 0xe006 of the factory partition (first and last byte of the address are enough, the rest can be removed if desired) and providing the MAC assignment on vendor firmware.

Reply:
`root@OpenWrt:~# hexdump -C /dev/mtd1
00000000  28 76 00 02 00 00 00 01  05 88 00 00 00 00 00 00  |(v..............|
00000010  ff ff ff ff ff ff ff ff  ff ff ff ff ff ff ff ff  |................|
00000020  00 00 00 00 20 00 00 00  00 0c 43 e1 76 29 00 0c  |.... .....C.v)..|
00000030  43 e1 76 2a 22 34 00 20  ff ff 00 01 00 00 00 00  |C.v*"4. ........|
00000040  00 00 22 00 00 00 00 00  30 00 00 00 00 00 00 00  |..".....0.......|
00000050  84 00 00 94 40 bb c0 ca  26 81 81 81 40 ca 26 80  |....@...&...@.&.|
00000060  80 80 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |................|
00000070  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |................|
*
000000a0  c1 c1 c1 c1 c1 c1 c0 c1  c1 c1 c1 c1 c1 c0 00 00  |................|
000000b0  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |................|
000000c0  00 00 00 00 00 00 00 00  00 1a 22 2a 31 35 01 35  |.........."*15.5|
000000d0  39 40 46 4d 7f 7f 7f 00  00 00 00 00 00 00 00 00  |9@FM............|
000000e0  11 1d 11 1d 1c 35 1c 35  1e 35 1e 35 17 19 17 19  |.....5.5.5.5....|
000000f0  02 00 00 00 b2 00 9c 88  0a 00 00 00 00 00 00 00  |................|
00000100  ff ff ff ff ff ff ff ff  ff ff ff ff ff ff ff ff  |................|
*
00000120  00 00 00 00 00 00 00 00  00 00 00 00 00 00 77 00  |..............w.|
00000130  11 1d 11 1d 15 7f 15 7f  17 7f 17 7f 10 3b 10 3b  |.............;.;|
00000140  ff ff ff ff ff ff ff ff  ff ff ff ff ff ff ff ff  |................|
*
00000200  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |................|
*
00001000  ff ff ff ff ff ff ff ff  ff ff ff ff ff ff ff ff  |................|
*
00008000  62 76 01 00 00 00 00 01  05 8a 12 76 c3 14 00 00  |bv.........v....|
00008010  00 00 12 76 c3 14 00 00  00 00 ff ff ff ff ff ff  |...v............|
00008020  ff ff 37 d8 9d 00 ff 7f  fd 9b ff ff ff ff ff ff  |..7.............|
00008030  ff ff ff ff 22 ff 00 20  ff ff b1 01 00 00 00 00  |....".. ........|
00008040  00 00 ff d7 00 00 00 00  00 00 00 00 00 00 e0 01  |................|
00008050  00 00 00 00 00 ae 7f 05  1a 00 00 00 7e 16 1a 00  |............~...|
00008060  00 00 82 e9 1a 00 00 82  e9 22 83 83 81 ed 22 83  |........."....".|
00008070  83 80 f0 22 83 83 81 f6  22 81 81 80 fe 22 81 81  |..."...."...."..|
00008080  82 e2 1a 00 00 82 e2 22  83 83 82 e8 22 83 83 82  |......."...."...|
00008090  e6 22 83 83 83 eb 22 83  83 83 f2 22 83 83 00 54  |."...."...."...T|
000080a0  c3 c3 c4 c4 00 00 c3 c3  c1 82 c3 c3 c1 82 00 00  |................|
000080b0  00 00 c3 c3 c3 00 00 00  00 00 00 00 00 00 85 85  |................|
000080c0  ff ff ff ff ff ff ff ff  ff ff ff ff ff ff ff ff  |................|
*
000080f0  ff ff ff ff ff ff a5 22  22 00 00 00 00 00 00 00  |......."".......|
00008100  0f ca 74 c5 e8 07 30 3d  01 b0 08 26 00 0e 04 15  |..t...0=...&....|
00008110  00 8a 00 40 00 00 00 08  00 9d 08 00 12 c0 00 00  |...@............|
00008120  08 20 04 2a 90 00 00 24  01 04 54 08 d0 a0 28 20  |. .*...$..T...( |
00008130  ff ff ff ff ff ff ff 08  ff ff ff ff ff ff 00 00  |................|
00008140  ff ff ff ff ff ff ff ff  ff ff ff ff ff ff ff ff  |................|
*
000081e0  c0 81 82 c3 04 45 46 07  08 09 ff ff ff ff ff ff  |.....EF.........|
000081f0  ff ff ff ff ff ff ff ff  ff ff ff ff ff ff ff ff  |................|
*
00020000
`
FW gives MACs:
LAN 00:00:00:01:05:88
WAN 00:00:00:01:05:89
WLAN2 00:00:00:01:05:88
WLAN5 00:00:00:01:05:8a

Reply:
I thought telling it to use code formatting would respect line breaks.  If that's too much of a mess I'll attach a txt file.

Reply:
I made the changes I proposed and checked OpenWRT's macs vs stock firmware and they match.  I've pushed the changes for your review.

Reply:
I fixed the formatting of your comment. Single backticks can only be used inline, not across lines. If you want to add a code block, you need triple backticks before and after the code.

Thanks for the data, this seems sane and looks correct in your implementation (except for the label MAC address issue). I'm just a little startled by the triple zero bytes. Please add your findings about MAC addresses on vendor FW and locations to the commit message, e.g. in a table like this:

```
MAC addresses based on vendor firmware:
LAN      *:88   0x4
WAN      *:89
WLAN2    *:88   0x4
WLAN5    *:8a   0x8004   (label)
```

Reply:
Unfortunately, the original contents of the factory partition appear to have been damaged (probably in one of my early flash attempts) and so the sticker mac is no longer present in memory.  When some of the data was still intact I found the sticker MAC + 4 in the 0x8004 address.  This is why I had the original strangeness for deriving the MAC, because I had apparently already damaged the data sitting at 0x4 before I got to this stage.  Unless someone else has an R6020 they can dump MTD1 out of, I'll have to procure another victim to confirm what the original values were.

As it stands, my configuration produces results that match what the OEM firmware outputs without resorting to any strange math or other tricks.  I removed the redundant alias line you pointed out and squashed again.

Reply:
Okay, I see. So even with possibly broken content, OEM firmware still told us the correct locations; and this is in line with the other devices as well.
I also think just removing label MAC address in this case is the best option, it can be easily added afterwards at any time (and it is only present for new devices anyway).

Reply:
Exactly what I was trying to convey.  Thank you.